model/riscv_insts_zfa.sail

```diff
@@ -685,54 +685,57 @@ function clause execute(RISCV_FLTQ_D(rs2, rs1, rd)) = {
+    let integer    = unsigned(fixedpoint[83..52]);
+    let fractional = to_bits(52, unsigned(fixedpoint[51..0]));
+
+    /* apply the sign bit */
+    let result = if (sign == 0b1) then negate(integer)
+                                  else integer;
+
+    /* raise FP exception flags, honoring the precedence of nV > nX */
+    let flags : bits(5) = if (true_exp > 31)         then nvFlag()
+                     else if (fractional != zeros()) then nxFlag()
+                     else                            zeros();
+
+    (flags, to_bits(32, result))
+}
+
 /*
  * Implement float64 to int32 conversion without saturation, the
  * result is supplied modulo 2^32.
- * Rounding mode for FCVTMOD.W.D is always RTZ(Round toward zero).
  */
-val      fcvtmod_helper : forall 'm, 'm == 64. (bits('m)) -> (bits(5), xlenbits)
+val      fcvtmod_helper : forall 'm, 'm == 64. (bits('m)) -> (bits(5), bits(32))
 function fcvtmod_helper(x64) = {
   let (sign, exp, mant) = fsplit_D(x64);
-  fflags		= fcsr.FFLAGS();
-  result : bits(64)	= undefined;
-
-  /* Handle special cases. */
-  if (exp == zeros())
-  then (nvFlag(), zeros())
-  /* NaN or Infinity. */
-  else if (exp == ones())
-  then (nvFlag(), zeros())
-  else {
-    let true_exp : int = unsigned(sub_vec_int(exp, 1023));
-    let shift    : int = true_exp - 52;
-    result	       = to_bits(64, unsigned(mant)) | ones() << 52;
-
-    /* Shift the fraction into place and set NX flag. */
-    if (shift >= 64 | shift <= -64) then {
-	/* The fraction is shifted out entirely. */
-	result = zeros();
-	fflags = fflags | nxFlag();
-    } else if (shift >= 0) then {
-	/* Raise NX if bit 52 got shifted out. */
-	result = result  << to_bits(64, shift);
-	if (unsigned(result) >= 12)
-	then fflags = fflags | nxFlag();
-    } else {
-	result = result >> to_bits(64, negate(shift));
-	/* Raise NX if bits got shifted out. */
-	if ((result << to_bits(64, negate(shift))) != result)
-	then fflags = fflags | nxFlag();
-    };
-
-    if (sign == 0b1)
-    then result = negate_D(result);
-
-    let result32b = to_bits(32, unsigned(result));
-    if (to_bits(64, unsigned(result32b)) != result)
-    then fflags = fflags | nvFlag();
-
-    (fflags, EXTS(result[31..0]))
-  };
+
+  /* detect the non-normal cases */
+  let is_subnorm    = (exp == zeros() & mant != zeros());
+  let is_zero       = (exp == zeros() & mant == zeros());
+  let is_nan_or_inf = (exp == ones());
+
+  /* For normal numbers, the mantissa and exponent use special encoding */
+  let true_mant     = 0b1 @ mant;  /* concatenate with the implicit bit */
+  let true_exp      = unsigned(exp) - 1023;   /* undo the offset-binary */
+
+  /* Detect the trivial cases for normal numbers */
+  let is_too_large  = true_exp >= 84;  /* bits will be 'multiplied' out */
+  let is_too_small  = true_exp <  0;   /* bits will be 'divided' out    */
+
+       if (is_zero)       then (zeros(),  zeros())
+  else if (is_subnorm)    then (nxFlag(), zeros())
```

Comment:
Could combine subnorm+too_small and nan_or_inf+too_large? At least I don't think that reordering changes semantics. In fact doesn't is_too_large imply is_nan_or_inf and is_subnorm+!is_zero imply is_too_small, i.e. you just need:
```
if (is_zero) then (zeros(), zeros())
else if (is_too_small) (nxFlag(), zeros())
else if (is_too_large) (nvFlag(), zeros())
else fcvtmod_helper_inner(sign, true_exp, true_mant)
```
unless I'm missing something here when glancing through it?

Reply:
While these can be combined, the interpretation of the exp and mantissa bits are different: e.g., for NaN and Inf, the mantissa bits are not interpreted as a mantissa (and the implicit bit doesn't play a role) ... similarly, exp is not decoded using binary-offset to form an exponent.

Given that this is an executable specification, we took the deliberate decision to keep these apart so the checks for the zero/subnorm case do not refer to decoded quantities that are (strictly speaking) not defined for the special cases.

I would expect a (in the far future — or possibly never) sufficiently optimizing compiler/interpreter to combine these cases when executing the specification.